File: 백준/Silver/10451. 순열 사이클/순열 사이클.py
import sys
sys.setrecursionlimit(10**6)  #재귀 깊이 제한을 늘리기위해
input = sys.stdin.readline

t = int(input())
def cycles(n,arr):
    visited = [False] * (n+1)
    cnt = 0
    def dfs(node):
        while not visited[node]:
            visited[node] = True
            node = arr[node]

    for i in range(1, n + 1):
        if not visited[i]:
            dfs(i)
            cnt += 1

    return cnt
for _ in range(t):
    n = int(input())
    arr =[0] + list(map(int,input().split()))
    print(cycles(n,arr))

File: 백준/Silver/10451. 순열 사이클/순열 사이클.java

import java.util.Scanner;
import java.util.StringTokenizer;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
public class Main {

     public static void main(String[] args) throws IOException {
          Scanner sc = new Scanner(System.in);
          int t = sc.nextInt();
          for(int i=0;i<t;i++){
               int n = sc.nextInt();
               int[] arr = new int[n+1];
               boolean[] visited = new boolean[n+1];

               for(int j = 1;j<=n;j++){
                    arr[j] = sc.nextInt();
               }
               int cnt = 0;
               for(int k = 1;k<=n;k++){
                    if(!visited[k]){
                         dfs(k,arr,visited);
                         cnt++;
                    }
               }
               System.out.println(cnt);
          }
          sc.close();
     }
     private static void dfs(int node, int[]arr, boolean[]visited){
          visited[node] = true;
          int nextnode = arr[node];
          if(!visited[nextnode]){
               dfs(nextnode,arr, visited);
          }
     }
}